### DISCUSSION TITLE
My vercel serverless fastapi code doesn't listen to startup events.

### DISCUSSION THREAD

**Discussion by JavaProgswing** [fastapi/fastapi#12614] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import logging

app = FastAPI()
logger = logging.getLogger()
db = None

@app.on_event("startup")
async def startup_event():
    global db
    db = "Connected"
    logger.info("Application started!")

@app.get("/")
async def read_root():
    global db
    logger.info(f"Database status: {db}")
    return {"message": "Hello, World!"}
```

### Description

The database doesn't get initialized because the startup event never gets called.

### Operating System

Other

### Operating System Details

Vercel

### FastAPI Version

0.115.3

### Pydantic Version

2.9.2

### Python Version

3.12

**Comment by dbfreem** [comment]
I think this is more of an issue with the logger config. If I change it to the below with basicConfig then I see the logs.

```python
from fastapi import FastAPI
import logging

app = FastAPI()
logging.basicConfig(level=logging.INFO)
db = None

@app.on_event("startup")
async def startup_event():
    global db
    db = "Connected"
    logging.info("Application started!")

@app.get("/")
async def read_root():
    global db
    logging.info(f"Database status: {db}")
    return {"message": "Hello, World!"}
```

```
INFO:     Finished server process [120898]
INFO:     Started server process [120910]
INFO:     Waiting for application startup.
INFO:root:Application started!
INFO:     Application startup complete.
INFO:watchfiles.main:3 changes detected
INFO:root:Database status: Connected
INFO:     127.0.0.1:43114 - "GET / HTTP/1.1" 200 OK

```

I noticed that the `app.on_event` was deprecated. You could use lifespan. Here is a modified version that I got from the docs [here](https://fastapi.tiangolo.com/advanced/events/?h=lifesp#lifespan)

```python
from contextlib import asynccontextmanager
from fastapi import FastAPI
import logging

logging.basicConfig(level=logging.DEBUG)

@asynccontextmanager
async def lifespan(app: FastAPI):
    global db
    db = "Connected"
    logging.info("Application started!")
    yield

app = FastAPI(lifespan=lifespan)

@app.get("/")
async def read_root():
    global db
    logging.info(f"Database status: {db}")
    return {"message": "Hello, World!"}
```


```
INFO:     Finished server process [121228]
INFO:     Started server process [121284]
INFO:     Waiting for application startup.
INFO:root:Application started!
INFO:     Application startup complete.
INFO:watchfiles.main:3 changes detected
INFO:root:Database status: Connected
INFO:     127.0.0.1:32824 - "GET / HTTP/1.1" 200 OK
```

**Reply by JavaProgswing** [reply]
How do you start fastapi? In vercel I see it being set to None instead of Connected.

**Reply by dbfreem** [reply]
`fastapi dev api/main.py`

**Reply by JavaProgswing** [reply]
Hmm vercel seems to do something strange, could this be related to it being a serverless deployment?

**Reply by dbfreem** [reply]
What startup command are you using in your vercel config?

**Comment by JavaProgswing** [comment]
I did not override the default startup command and I cannot determine what
was the default startup command.
